### PR TITLE
Detect circular config and skip

### PR DESCRIPTION
**Pull request by pooja-postman**
[Sails equivalent of the photon PR#156](https://github.com/postman-eng/photon/pull/156/files)

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -20,7 +17,7 @@ var E = '',
      * @returns {string}
      */
     rekey = function (key, separator) {
-        return (typeof key === NUMBER) ? key : (key.toUpperCase().replace(/[^A-Z0-9]/g, separator || E));
+        return key.toUpperCase().replace(/[^A-Z0-9]/g, separator || E);
```
Comment (agrawalmehul) [inline lib/main.js]: Why did we remove the `typeof key === NUMBER` check?

File: lib/main.js [hunk 2]
```diff
@@ -60,29 +57,33 @@ var E = '',
  * @param {string=} [separator=_]
  * @param {function} transformer - The key name to pick data is forwarded as parameter and the value returned by the
  * transformer function is used to lookup environment variables.
+ * @param {boolean} _debug
```
Comment (agrawalmehul) [inline lib/main.js:62]: Needs `visitedObjects` JSDoc entry

File: lib/main.js [hunk 3]
```diff
@@ -60,29 +57,33 @@ var E = '',
...
-lift = function (obj, namespace, separator, transformer, _debug) {
-    var cast,
-        eKey,
-        allKeys;
+lift = function (obj, namespace, separator, transformer, _debug, visitedObjects) {
+    var key,
+        cast,
+        eKey;
 
-    // iterate on each item in this object and operate on their keys.
```
Comment (agrawalmehul) [inline lib/main.js:69]: Did we remove this entire block of logic?

File: tests/unit/examples-spec.js [hunk 4]
```diff
@@ -36,4 +41,10 @@ describe('examples', function () {
             port: 2222
         }]);
     });
+
+    it('same name sub-property as parent property', function () {
```
Comment (agrawalmehul) [inline tests/unit/examples-spec.js:49]: Should also check that winstonLogger property is left as is?

Comment (pooja-postman) [reply tests/unit/examples-spec.js:49]: Done above in `sails-local` config

File: lib/main.js [hunk 5]
```diff
@@ -79,9 +94,11 @@ lift = function (obj, namespace, separator, transformer, _debug) {
         allKeys = [];  // If this happens, the loop will be skipped completely
     }
 
+    // iterate on each item in this object and operate on their keys.
     allKeys.forEach(function (key) {
+        // Photon
```
Comment (agrawalmehul) [inline lib/main.js]: `Photon` mention on public package.

File: lib/main.js [hunk 6]
```diff
@@ -60,11 +59,27 @@ var E = '',
...
-lift = function (obj, namespace, separator, transformer, _debug) {
-    var cast,
-        eKey,
-        allKeys;
+lift = function (obj, namespace, separator, transformer, _debug, visitedObjects) {
+    var allKeys,
+        cast,
+        eKey;
+
+    // Photon - not loading function values from environment config
```
Comment (agrawalmehul) [inline lib/main.js]: Photon reference

Comment (agrawalmehul) [reply lib/main.js]: same on line 75

**Review by agrawalmehul — APPROVED** [review]

**Review by agrawalmehul — APPROVED** [review]

**Review by agrawalmehul — APPROVED** [review]

**Pull request merged**